File: quant/portfolio/alpha.py
"""Alpha combination and ensemble framework.

Combines multiple signal scores into a single composite alpha per symbol,
supporting static weights, inverse-volatility weighting, and rank-based
combination.

Scoring convention (inherited from signals):
  +1.0 = maximum long conviction
  -1.0 = maximum short conviction
   0.0 = flat / no view
"""
from __future__ import annotations

import enum
from dataclasses import dataclass, field
from datetime import datetime

from quant.signals.base import SignalOutput


@dataclass(frozen=True, slots=True)
class AlphaScore:
    """Composite alpha for a single symbol at a point in time.

    Attributes:
        symbol:     Ticker symbol.
        timestamp:  Bar timestamp.
        score:      Combined alpha in [-1, 1].
        confidence: Aggregate confidence in [0, 1].
        signal_contributions: Per-signal weighted contribution to the final score.
    """

    symbol: str
    timestamp: datetime
    score: float
    confidence: float
    signal_contributions: dict[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not (-1.0 <= self.score <= 1.0):
            raise ValueError(f"score must be in [-1, 1], got {self.score}")
        if not (0.0 <= self.confidence <= 1.0):
            raise ValueError(f"confidence must be in [0, 1], got {self.confidence}")


class CombinationMethod(enum.Enum):
    STATIC_WEIGHT = "static_weight"
    INVERSE_VOLATILITY = "inverse_volatility"
    RANK_WEIGHTED = "rank_weighted"
    EQUAL_WEIGHT = "equal_weight"


def _clamp(value: float, lo: float = -1.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, value))


class AlphaCombiner:
    """Combines multiple signal outputs into a single composite alpha.

    Usage::

        combiner = AlphaCombiner(
            method=CombinationMethod.STATIC_WEIGHT,
            weights={"momentum": 0.4, "mean_reversion": 0.3, "trend_following": 0.3},
        )
        alpha = combiner.combine("AAPL", timestamp, signal_outputs)
    """

    def __init__(
        self,
        method: CombinationMethod = CombinationMethod.EQUAL_WEIGHT,
        weights: dict[str, float] | None = None,
    ) -> None:
        self._method = method
        self._weights = weights or {}
        if method == CombinationMethod.STATIC_WEIGHT and not self._weights:
            raise ValueError("weights required for STATIC_WEIGHT method")

    @property
    def method(self) -> CombinationMethod:
        return self._method

    def combine(
        self,
        symbol: str,
        timestamp: datetime,
        signals: list[SignalOutput],
    ) -> AlphaScore:
        """Combine signal outputs into a composite alpha score.

        Args:
            symbol:    Ticker symbol.
            timestamp: Current bar timestamp.
            signals:   List of SignalOutput from different strategies.

        Returns:
            AlphaScore with combined score and per-signal contributions.
        """
        if not signals:
            return AlphaScore(
                symbol=symbol,
                timestamp=timestamp,
                score=0.0,
                confidence=0.0,
            )

        if self._method == CombinationMethod.STATIC_WEIGHT:
            return self._static_weight(symbol, timestamp, signals)
        if self._method == CombinationMethod.INVERSE_VOLATILITY:
            return self._inverse_volatility(symbol, timestamp, signals)
        if self._method == CombinationMethod.RANK_WEIGHTED:
            return self._rank_weighted(symbol, timestamp, signals)
        return self._equal_weight(symbol, timestamp, signals)

    def combine_universe(
        self,
        timestamp: datetime,
        universe_signals: dict[str, list[SignalOutput]],
    ) -> dict[str, AlphaScore]:
        """Combine signals for every symbol in the universe.

        Args:
            timestamp:        Current bar timestamp.
            universe_signals: Dict of {symbol: [SignalOutput, ...]}.

        Returns:
            Dict of {symbol: AlphaScore}.
        """
        return {
            sym: self.combine(sym, timestamp, sigs)
            for sym, sigs in universe_signals.items()
        }

    def _static_weight(
        self, symbol: str, timestamp: datetime, signals: list[SignalOutput]
    ) -> AlphaScore:
        contributions: dict[str, float] = {}
        total_weight = 0.0
        weighted_score = 0.0
        weighted_conf = 0.0

        for sig in signals:
            # Look up weight by signal name from metadata, or default to 0
            sig_name = sig.metadata.get("signal_name", "")
            w = self._weights.get(sig_name, 0.0)
            contributions[sig_name] = sig.score * w
            weighted_score += sig.score * w
            weighted_conf += sig.confidence * w
            total_weight += w

        if total_weight > 0:
            weighted_conf /= total_weight

        return AlphaScore(
            symbol=symbol,
            timestamp=timestamp,
            score=_clamp(weighted_score),
            confidence=_clamp(weighted_conf, 0.0, 1.0),
            signal_contributions=contributions,
        )

    def _equal_weight(
        self, symbol: str, timestamp: datetime, signals: list[SignalOutput]
    ) -> AlphaScore:
        n = len(signals)
        contributions: dict[str, float] = {}
        total_score = 0.0
        total_conf = 0.0

        for sig in signals:
            sig_name = sig.metadata.get("signal_name", f"signal_{id(sig)}")
            contribution = sig.score / n
            contributions[sig_name] = contribution
            total_score += contribution
            total_conf += sig.confidence

        return AlphaScore(
            symbol=symbol,
            timestamp=timestamp,
            score=_clamp(total_score),
            confidence=_clamp(total_conf / n, 0.0, 1.0),
            signal_contributions=contributions,
        )

    def _inverse_volatility(
        self, symbol: str, timestamp: datetime, signals: list[SignalOutput]
    ) -> AlphaScore:
        """Weight signals inversely to their confidence uncertainty (1 - confidence).

        Higher-confidence signals receive proportionally more weight.
        """
        contributions: dict[str, float] = {}
        # Use confidence as a proxy for signal quality
        raw_weights: list[float] = []
        for sig in signals:
            # inv_vol weight: higher confidence → lower "vol" → higher weight
            raw_weights.append(sig.confidence + 1e-8)

        total_w = sum(raw_weights)
        weights = [w / total_w for w in raw_weights]

        weighted_score = 0.0
        weighted_conf = 0.0
        for sig, w in zip(signals, weights, strict=True):
            sig_name = sig.metadata.get("signal_name", f"signal_{id(sig)}")
            contributions[sig_name] = sig.score * w
            weighted_score += sig.score * w
            weighted_conf += sig.confidence * w

        return AlphaScore(
            symbol=symbol,
            timestamp=timestamp,
            score=_clamp(weighted_score),
            confidence=_clamp(weighted_conf, 0.0, 1.0),
            signal_contributions=contributions,
        )

    def _rank_weighted(
        self, symbol: str, timestamp: datetime, signals: list[SignalOutput]
    ) -> AlphaScore:
        """Rank signals by absolute score magnitude; higher-ranked get more weight."""
        contributions: dict[str, float] = {}
        # Sort by |score| descending
        sorted_sigs = sorted(signals, key=lambda s: abs(s.score), reverse=True)
        n = len(sorted_sigs)

        # Rank weights: top signal gets n, second gets n-1, etc.
        raw_weights = [float(n - i) for i in range(n)]
        total_w = sum(raw_weights)
        weights = [w / total_w for w in raw_weights]

        weighted_score = 0.0
        weighted_conf = 0.0
        for sig, w in zip(sorted_sigs, weights, strict=True):
            sig_name = sig.metadata.get("signal_name", f"signal_{id(sig)}")
            contributions[sig_name] = sig.score * w
            weighted_score += sig.score * w
            weighted_conf += sig.confidence * w

        return AlphaScore(
            symbol=symbol,
            timestamp=timestamp,
            score=_clamp(weighted_score),
            confidence=_clamp(weighted_conf, 0.0, 1.0),
            signal_contributions=contributions,
        )

File: quant-rs/quant-portfolio/src/alpha.rs
//! Alpha score types and combination methods.
//!
//! Mirrors `quant.portfolio.alpha`.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

// ── Alpha score ───────────────────────────────────────────────────────────────

/// A composite alpha score for a single asset.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlphaScore {
    pub symbol: String,
    /// Combined signal score in [-1, 1].
    pub score: f64,
    /// Confidence in [0, 1].
    pub confidence: f64,
    /// Target position fraction (score × confidence), clamped to [-1, 1].
    pub target_position: f64,
}

impl AlphaScore {
    pub fn new(symbol: impl Into<String>, score: f64, confidence: f64) -> Self {
        let target_position = (score * confidence).clamp(-1.0, 1.0);
        Self {
            symbol: symbol.into(),
            score: score.clamp(-1.0, 1.0),
            confidence: confidence.clamp(0.0, 1.0),
            target_position,
        }
    }
}

// ── Combination method ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CombinationMethod {
    /// Simple average of all signal scores.
    EqualWeight,
    /// Weighted average using explicit per-signal weights.
    StaticWeight,
    /// Weight proportional to confidence × absolute score.
    ConvictionWeighted,
}

// ── Individual signal input ───────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct SignalInput {
    pub signal_name: String,
    pub score: f64,
    pub confidence: f64,
    pub target_position: f64,
}

// ── AlphaCombiner ─────────────────────────────────────────────────────────────

/// Combines multiple signal outputs into a single alpha score per asset.
pub struct AlphaCombiner {
    pub method: CombinationMethod,
    /// Static signal weights (used with `StaticWeight`).
    pub weights: Option<HashMap<String, f64>>,
}

impl AlphaCombiner {
    pub fn new(method: CombinationMethod, weights: Option<HashMap<String, f64>>) -> Self {
        Self { method, weights }
    }

    /// Combine a list of signal inputs for a single symbol.
    pub fn combine(&self, symbol: &str, signals: &[SignalInput]) -> AlphaScore {
        if signals.is_empty() {
            return AlphaScore::new(symbol, 0.0, 0.0);
        }

        let (score, confidence) = match self.method {
            CombinationMethod::EqualWeight => self.combine_equal(signals),
            CombinationMethod::StaticWeight => self.combine_static(signals),
            CombinationMethod::ConvictionWeighted => self.combine_conviction(signals),
        };

        AlphaScore::new(symbol, score, confidence)
    }

    fn combine_equal(&self, signals: &[SignalInput]) -> (f64, f64) {
        let n = signals.len() as f64;
        let score = signals.iter().map(|s| s.score).sum::<f64>() / n;
        let confidence = signals.iter().map(|s| s.confidence).sum::<f64>() / n;
        (score, confidence)
    }

    fn combine_static(&self, signals: &[SignalInput]) -> (f64, f64) {
        let weights = match &self.weights {
            Some(w) => w,
            None => return self.combine_equal(signals),
        };

        let mut total_weight = 0.0_f64;
        let mut score_sum = 0.0_f64;
        let mut conf_sum = 0.0_f64;

        for s in signals {
            let w = weights.get(&s.signal_name).copied().unwrap_or(1.0);
            score_sum += s.score * w;
            conf_sum += s.confidence * w;
            total_weight += w;
        }

        if total_weight == 0.0 {
            return (0.0, 0.0);
        }
        (score_sum / total_weight, conf_sum / total_weight)
    }

    fn combine_conviction(&self, signals: &[SignalInput]) -> (f64, f64) {
        let mut total_weight = 0.0_f64;
        let mut score_sum = 0.0_f64;
        let mut conf_sum = 0.0_f64;

        for s in signals {
            let w = (s.confidence * s.score.abs()).max(1e-9);
            score_sum += s.score * w;
            conf_sum += s.confidence * w;
            total_weight += w;
        }

        if total_weight == 0.0 {
            return self.combine_equal(signals);
        }
        (score_sum / total_weight, conf_sum / total_weight)
    }
}

// ── Unit tests ────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn signals() -> Vec<SignalInput> {
        vec![
            SignalInput { signal_name: "momentum".into(), score: 0.8, confidence: 0.6, target_position: 0.48 },
            SignalInput { signal_name: "mean_reversion".into(), score: -0.4, confidence: 0.9, target_position: -0.36 },
        ]
    }

    #[test]
    fn test_equal_weight_combine() {
        let combiner = AlphaCombiner::new(CombinationMethod::EqualWeight, None);
        let alpha = combiner.combine("AAPL", &signals());
        assert!((alpha.score - 0.2).abs() < 1e-9);
        assert!((alpha.confidence - 0.75).abs() < 1e-9);
    }

    #[test]
    fn test_empty_signals_returns_zero() {
        let combiner = AlphaCombiner::new(CombinationMethod::EqualWeight, None);
        let alpha = combiner.combine("AAPL", &[]);
        assert!(alpha.score.abs() < 1e-12);
        assert!(alpha.confidence.abs() < 1e-12);
    }

    #[test]
    fn test_alpha_score_clamps() {
        let a = AlphaScore::new("X", 2.0, 1.5);
        assert_eq!(a.score, 1.0);
        assert_eq!(a.confidence, 1.0);
    }
}
